src/control/server/mgmt_svc.go

```diff
@@ -198,9 +215,27 @@ func (svc *mgmtSvc) Join(ctx context.Context, req *mgmtpb.JoinReq) (*mgmtpb.Join
 
 	// if join successful, record membership
 	if resp.GetStatus() == 0 && resp.GetState() == mgmtpb.JoinResp_IN {
-		svc.members.Add(common.SystemMember{
-			Addr: req.GetAddr(), Uuid: req.GetUuid(), Rank: resp.GetRank(),
-		})
+		newMember := common.SystemMember{Uuid: req.GetUuid(), Rank: resp.GetRank()}
+
+		reqAddr, err := net.ResolveTCPAddr("tcp", req.GetAddr())
+		if err != nil {
+			// leave newMember.Addr uninitialised
+			svc.log.Errorf("resolving request address: %s", err)
+			svc.members.Add(newMember)
```

Comment:
Also here... If the goal is to build up a membership set which is used to figure out who should get RPCs, adding members with unresolvable addresses seems like it will lead to problems later.

Reply:
I was struggling to work out the situation, so what should happen if a join occurs but we can't resolve the address?

Reply:
It's worth looking at what could lead to a join with an unresolvable address. If it's a "can't happen" kind of situation, then I'd say we should log it at error and reject it. If there is a valid use case for allowing a join with a bad address then that's a different story, but it should be fleshed out.

My inclination would be to log/reject as a first pass and deal with handling members with bad addresses later if it comes up that we need to support that somehow. Otherwise we risk getting into weird situations without consciously choosing to go there...

Reply:
I should imagine that occasional address resolution failures due to temporal network issues should be an expected failure case. happy to deal with that at a later date and I think there is even a address error field that reflects a temporary failure type https://golang.org/src/net/net.go?s=17039:17075#L545 that we might be able to leverage in the future. 

For the moment I'm fine to go with log/reject as we are at an early stage.

Reply:
and some point we may have to deal with different types of addressing such as from proxies where we won't have the expected details in peer and will have to do some lookups

Reply:
Yep, I agree that it's probably going to happen. The question is: What should we do about it? Adding a member with an empty address seems wrong -- it's kicking the can down the road for some other part of the codebase to deal with (perhaps multiple times). Seems better to deal with it once, in one place.

Looking at the bigger picture, I notice that the peer address resolution is being done after the dRPC join call. So we may have a situation where the member has been added to the management service but we're not recording it in the control plane's membership list.

I suggest moving the address resolution step before calling dRPC join. If we can't resolve the member's address, it may indicate a bigger problem and it would probably be best to fail quickly and noisily.